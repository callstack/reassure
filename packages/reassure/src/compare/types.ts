export const STATISTICAL_SIGNIFICANCE = ['SIGNIFICANT', 'INSIGNIFICANT', 'MEANINGLESS'] as const;

/**
 * Entry in the performance results file.
 */
export interface PerformanceEntry {
  /** Name of the test scenario */
  name: string;

  /** Number of times the measurment test was run */
  runs: number;

  /** Array of measured render durations for each run */
  durations: number[];

  /** Arithmetic average of measured render durations for each run */
  meanDuration: number;

  /* Standard deviation of measured render durations for each run */
  stdevDuration: number;

  /** Array of measured render counts for each run */
  counts: number[];

  /** Arithmetic average of measured render counts for each run */
  meanCount: number;

  /** Standard deviation of measured render counts for each run */
  stdevCount: number;
}

/**
 * Type of the performance measure change as compared to the baseline results
 */
export type StatisticalSignificance = typeof STATISTICAL_SIGNIFICANCE[number];

/**
 * Compare entry for tests that have both baseline and current entry
 */
export interface CompareEntry {
  name: string;
  current: PerformanceEntry;
  baseline: PerformanceEntry;
  durationDiff: number;
  relativeDurationDiff: number;
  durationDiffSignificance: StatisticalSignificance;
  countDiff: number;
  relativeCountDiff: number;
}

/**
 * Compare entry for tests that have only current entry
 */
export interface AddedEntry {
  name: string;
  current: PerformanceEntry;
}

/**
 * Compare entry for tests that have only baseline entry
 */
export interface RemovedEntry {
  name: string;
  baseline: PerformanceEntry;
}

/**
 * Output of compare function
 */
export type CompareResult = {
  significant: CompareEntry[];
  insignificant: CompareEntry[];
  meaningless: CompareEntry[];
  countChanged: CompareEntry[];
  added: AddedEntry[];
  removed: RemovedEntry[];
  errors: string[];
  warnings: string[];
};
