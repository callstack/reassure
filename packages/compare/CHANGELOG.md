# @callstack/reassure-compare

## 1.0.0-rc.5

### Minor Changes

- 04be1d4: Detect render issues (initial render updates, redundant renders)

### Patch Changes

- @callstack/reassure-logger@1.0.0-rc.5

## 1.0.0-rc.4

### Patch Changes

- @callstack/reassure-logger@1.0.0-rc.4

## 1.0.0-rc.3

### Minor Changes

- 2ef5f4d: chore: fix version deps

### Patch Changes

- Updated dependencies [2ef5f4d]
  - @callstack/reassure-logger@1.0.0-rc.3

## 1.0.0-rc.2

### Minor Changes

- f722a6d: chore: migrate to Yarn Berry (4.x)

### Patch Changes

- Updated dependencies [f722a6d]
  - @callstack/reassure-logger@1.0.0-rc.2

## 1.0.0 RC

### Major Changes

- v1.0

### Patch Changes

- Updated dependencies [f9116d0]
  - @callstack/reassure-logger@1.0.0-rc.1

## 0.11.0

### Patch Changes

- b455bd4: refactor: simplify `reassure-logger` package exports
- bb2046a: Replace unmaintained `markdown-builder` library with internal helper functions.
- Updated dependencies [b455bd4]
  - @callstack/reassure-logger@0.11.0

## 0.6.0

### Minor Changes

- feat: `measureFunction` API to measure regular JS functions execution time

### Patch Changes

- Updated dependencies [99afdf98]
  - @callstack/reassure-logger@0.3.2

## 0.5.1

### Patch Changes

- Updated dependencies [2e127815]
  - @callstack/reassure-logger@0.3.1

## 0.5.0

### Minor Changes

- f8aa25fe: feat: capture date and time of measurements

## 0.4.1

### Patch Changes

- Updated dependencies [235f37d4]
  - @callstack/reassure-logger@0.3.0

## 0.4.0

### Minor Changes

- 35af62a4: Custom logger implementation with verbose and silent modes.

### Patch Changes

- Updated dependencies [35af62a4]
  - @callstack/reassure-logger@0.2.0

## 0.3.0

### Minor Changes

- 3a180f2: Validate format of performance results files when loading for comparison

## 0.2.0

### Minor Changes

- b4250e3c: Include codebase metadata (branch, commit hash) in the performance report and measurement file

### Patch Changes

- 5a1c3472: Changes for dependencies cleanup after monorepo migration

## 0.1.1

### Patch Changes

- 4d0cca6a: Updated console output format
- ca56a6d1: Move internal packages under @reassure scope to @callstack scope for simpler maintenance

## 0.1.0

### Minor Changes

- d6bfef03: Add note about ignoring .reassure folder

## 0.0.3

### Patch Changes

- 27f45e83: Fix typescript setup for publishing with bob

## 0.0.2

### Patch Changes

- 2f8f8c06: setup changesets
