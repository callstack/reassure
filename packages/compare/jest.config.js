module.exports = {
  preset: '@testing-library/react-native',
  transformIgnorePatterns: ['node_modules/(?!(jest-)?@?react-native|@react-native-community|@react-navigation)'],
  modulePathIgnorePatterns: ['<rootDir>/lib/'],
  snapshotSerializers: ['@relmify/jest-serializer-strip-ansi/always'],
  clearMocks: true,
};
