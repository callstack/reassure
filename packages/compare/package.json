{
  "name": "@callstack/reassure-compare",
  "version": "1.4.0",
  "description": "Performance testing companion for React and React Native",
  "main": "lib/commonjs/index.js",
  "module": "lib/module/index.js",
  "types": "lib/typescript/index.d.ts",
  "source": "src/index.ts",
  "files": [
    "lib",
    "!**/__tests__",
    "!**/__fixtures__",
    "!**/__mocks__"
  ],
  "scripts": {
    "build": "bob build",
    "clean": "del lib",
    "test": "jest"
  },
  "keywords": [
    "react-native",
    "ios",
    "android"
  ],
  "repository": "https://github.com/callstack/reassure",
  "author": "Maciej Jastrzębski <mdjastrzebski@gmail.com> (https://github.com/mdjastrzebski)",
  "contributors": [
    "Jakub Bujko <jakub.bujko@callstack.com> (https://github.com/Xiltyn)",
    "Tomasz Krzyżowski <tomasz.krzyzowski@callstack.com> (https://github.com/TMaszko)",
    "Michał Pierzchała <thymikee@gmail.com> (https://github.com/thymikee)"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/callstack/reassure/issues"
  },
  "homepage": "https://github.com/callstack/reassure#readme",
  "dependencies": {
    "@callstack/reassure-logger": "1.4.0",
    "ts-markdown-builder": "0.4.1",
    "ts-regex-builder": "^1.8.2",
    "zod": "^3.24.2"
  },
  "devDependencies": {
    "@babel/core": "^7.26.9",
    "@babel/plugin-transform-flow-strip-types": "^7.26.5",
    "@babel/preset-env": "^7.26.9",
    "@babel/preset-typescript": "^7.26.0",
    "@babel/runtime": "^7.26.9",
    "@relmify/jest-serializer-strip-ansi": "^1.0.2",
    "@types/jest": "^30.0.0",
    "@types/react": "^19.0.0",
    "babel-jest": "^30.0.2",
    "del-cli": "^5.1.0",
    "jest": "^30.0.2",
    "prettier": "^2.8.8",
    "react-native-builder-bob": "^0.24.0",
    "typescript": "^5.8.2"
  },
  "react-native-builder-bob": {
    "source": "src",
    "output": "lib",
    "targets": [
      "commonjs",
      "module",
      [
        "typescript",
        {
          "project": "tsconfig.build.json"
        }
      ]
    ]
  },
  "publishConfig": {
    "access": "public"
  }
}
