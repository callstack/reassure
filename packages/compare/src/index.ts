export { compare } from './compare';
export { formatMetadata } from './utils/format';

export type {
  MeasureHeader,
  MeasureMetadata,
  MeasureEntry,
  CompareResult,
  CompareMetadata,
  CompareEntry,
  AddedEntry,
  RemovedEntry,
  RenderIssues,
} from './types';
