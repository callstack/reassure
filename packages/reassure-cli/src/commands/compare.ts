import type { CommandModule } from 'yargs';
import { compare } from '@reassure/reassure-compare';

export async function run() {
  return compare({});
}

export const command: CommandModule = {
  command: 'compare',
  describe: 'Compares gathered baseline and current measurements',
  handler: () => run(),
};
