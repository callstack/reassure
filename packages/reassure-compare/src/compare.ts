import * as fsSync from 'fs';

import type {
  AddedEntry,
  RemovedEntry,
  CompareEntry,
  CompareResult,
  PerformanceResults,
  PerformanceEntry,
  PerformanceHeader,
} from './types';
import { printToConsole } from './output/console';
import { writeToJson } from './output/json';
import { writeToMarkdown } from './output/markdown';
import { errors, warnings, logError, logWarning } from './utils/logs';
import { parseHeader, parsePerformanceEntries } from './utils/validate';

/**
 * Probability threshold for considering given difference significant.
 */
const PROBABILITY_CONSIDERED_SIGNIFICANT = 0.02;

/**
 * Render duration threshold (in ms) for treating given difference as significant.
 *
 * This is additional filter, in addition to probability threshold above.
 * Too small duration difference might be result of measurement grain of 1 ms.
 */
const DURATION_DIFF_THRESHOLD_SIGNIFICANT = 4;

/**
 * Threshold for considering render count change as significant. This implies inclusion
 * of scenario results in Render Count Changed output section.
 */
const COUNT_DIFF_THRESHOLD = 0.5;

type CompareOptions = {
  baselineFile?: string;
  currentFile?: string;
  outputFile?: string;
  outputFormat?: string;
};

/**
 * Main routine.
 *
 * Responsible for loading baseline and current performance results and outputting data in various formats.
 */
export function compare({
  baselineFile = '.reassure/baseline.perf',
  currentFile = '.reassure/current.perf',
  outputFile = '.reassure/output.json',
  outputFormat = 'all',
}: CompareOptions = {}) {
  const hasCurrentFile = fsSync.existsSync(currentFile);
  if (!hasCurrentFile) {
    logError(`Current results files "${currentFile}" does not exists. Check your setup.`);
    process.exit(1);
  }
  let currentResults: PerformanceResults;
  try {
    currentResults = loadFile(currentFile);
  } catch (error) {
    logError(`Error while parsing file: ${currentFile}`, error);
    process.exit(1);
  }

  const hasBaselineFile = fsSync.existsSync(baselineFile);
  if (!hasBaselineFile) {
    logWarning(
      `Baseline results files "${baselineFile}" does not exists. This warning should be ignored only if you are bootstapping perf test setup, otherwise it indicates invalid setup.`
    );
  }

  let baselineResults: PerformanceResults | null = null;
  if (hasBaselineFile) {
    try {
      baselineResults = loadFile(baselineFile);
    } catch (error) {
      logError(`Error while parsing file: ${baselineFile}`, error);
      process.exit(1);
    }
  }

  const output = compareResults(currentResults, baselineResults);

  if (outputFormat === 'console' || outputFormat === 'all') {
    printToConsole(output);
  }
  if (outputFormat === 'json' || outputFormat === 'all') {
    writeToJson(outputFile, output);
  }
  if (outputFormat === 'markdown' || outputFormat === 'all') {
    writeToMarkdown('.reassure/output.md', output);
  }
}

/**
 * Load performance file and parse it to `PerformanceRecord` object.
 */
export function loadFile(path: string): PerformanceResults {
  const contents = fsSync.readFileSync(path, 'utf8');

  const lines = contents
    .split(/\r?\n/)
    .filter((line) => !!line.trim())
    .map((line) => JSON.parse(line));

  let header: PerformanceHeader | null = null;
  let entries: PerformanceEntry[] = [];

  const hasHeader = lines[0].metadata !== undefined;
  if (hasHeader) {
    header = parseHeader(lines[0]);
    entries = parsePerformanceEntries(lines.slice(1));
  } else {
    entries = parsePerformanceEntries(lines);
  }

  const keyedEntries: Record<string, PerformanceEntry> = {};
  entries.forEach((entry) => {
    keyedEntries[entry.name] = entry;
  });

  return {
    metadata: header?.metadata,
    entries: keyedEntries,
  };
}

/**
 * Compare results between baseline and current entries and categorize.
 */
function compareResults(current: PerformanceResults, baseline: PerformanceResults | null): CompareResult {
  // Unique test scenario names
  const names = [...new Set([...Object.keys(current.entries), ...Object.keys(baseline?.entries || {})])];
  const compared: CompareEntry[] = [];
  const added: AddedEntry[] = [];
  const removed: RemovedEntry[] = [];

  names.forEach((name) => {
    const currentEntry = current.entries[name];
    const baselineEntry = baseline?.entries[name];

    if (currentEntry && baselineEntry) {
      compared.push(buildCompareEntry(name, currentEntry, baselineEntry));
    } else if (currentEntry) {
      added.push({ name, current: currentEntry });
    } else if (baselineEntry) {
      removed.push({ name, baseline: baselineEntry });
    }
  });

  const significant = compared
    .filter((item) => item.isDurationDiffSignificant)
    .sort((a, b) => b.durationDiff - a.durationDiff);
  const meaningless = compared
    .filter((item) => !item.isDurationDiffSignificant)
    .sort((a, b) => b.durationDiff - a.durationDiff);
  const countChanged = compared
    .filter((item) => Math.abs(item.countDiff) > COUNT_DIFF_THRESHOLD)
    .sort((a, b) => b.countDiff - a.countDiff);
  added.sort((a, b) => b.current.meanDuration - a.current.meanDuration);
  removed.sort((a, b) => b.baseline.meanDuration - a.baseline.meanDuration);

  return {
    metadata: { current: current.metadata, baseline: baseline?.metadata },
    errors,
    warnings,
    significant,
    meaningless,
    countChanged,
    added,
    removed,
  };
}

/**
 * Establish statisticial significance of render duration difference build compare entry.
 */
function buildCompareEntry(name: string, current: PerformanceEntry, baseline: PerformanceEntry): CompareEntry {
  const durationDiff = current.meanDuration - baseline.meanDuration;
  const relativeDurationDiff = durationDiff / baseline.meanDuration;
  const countDiff = current.meanCount - baseline.meanCount;
  const relativeCountDiff = countDiff / baseline.meanCount;

  const z = computeZ(baseline.meanDuration, baseline.stdevDuration, current.meanDuration, current.runs);
  const prob = computeProbability(z);

  const isDurationDiffSignificant =
    prob < PROBABILITY_CONSIDERED_SIGNIFICANT && Math.abs(durationDiff) >= DURATION_DIFF_THRESHOLD_SIGNIFICANT;

  return {
    name,
    baseline,
    current,
    durationDiff,
    relativeDurationDiff,
    isDurationDiffSignificant,
    countDiff,
    relativeCountDiff,
  };
}

/**
 * Calculate z-score for given baseline and current performance results.
 *
 * Based on :: https://github.com/v8/v8/blob/master/test/benchmarks/csuite/compare-baseline.py
 */
function computeZ(baselineMean: number, baselineStdev: number, currentMean: number, runs: number) {
  if (baselineStdev == 0) return 1000;

  return Math.abs((currentMean - baselineMean) / (baselineStdev / Math.sqrt(runs)));
}

/**
 * Compute statisticial hyphotesis probability based on z-score.
 *
 * Based on :: https://github.com/v8/v8/blob/master/test/benchmarks/csuite/compare-baseline.py
 */
function computeProbability(z: number) {
  // p 0.005: two sided < 0.01
  if (z > 2.575_829) return 0;

  // p 0.010
  if (z > 2.326_348) return 0.01;

  // p 0.015
  if (z > 2.170_091) return 0.02;

  // p 0.020
  if (z > 2.053_749) return 0.03;

  // p 0.025: two sided < 0.05
  if (z > 1.959_964) return 0.04;

  // p 0.030
  if (z > 1.880_793) return 0.05;

  // p 0.035
  if (z > 1.811_91) return 0.06;

  // p 0.040
  if (z > 1.750_686) return 0.07;

  // p 0.045
  if (z > 1.695_397) return 0.08;

  // p 0.050: two sided < 0.10
  if (z > 1.644_853) return 0.09;

  // p 0.100: two sided < 0.20
  if (z > 1.281_551) return 0.1;

  // two sided p >= 0.20
  return 0.2;
}
