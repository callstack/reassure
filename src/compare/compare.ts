import * as fs from 'fs/promises';
import * as fsSync from 'fs';
import minimist from 'minimist';

import type { PerformanceEntry } from '../measure/types';
import { hasDuplicateValues } from '../utils/array';
import type { StatisticalSignificance, AddedEntry, RemovedEntry, CompareEntry, CompareResult } from './types';
import { printToConsole } from './outputConsole';
import { writeToJson } from './outputJson';
import { writeToMarkdown } from './outputMarkdown';

type ScriptArguments = {
  baselineFilePath: string;
  currentFilePath: string;
  outputFilePath: string;
  output?: 'console' | 'json' | 'markdown' | 'all';
};

/**
 * Record type holding `PerformanceEntry` objects keyed by entry name.
 */
type PerformanceRecord = { [name: string]: PerformanceEntry };

/**
 * List of arguments which can be passed to the node command running the script as --<ARGUMENT>=<VALUE>
 * e.g. --baselineFilePath="./myOutputFile.txt"
 */
const {
  output,
  currentFilePath = 'perf-results.txt',
  baselineFilePath = 'baseline-results.txt',
  outputFilePath = 'compare-output.json',
} = minimist<ScriptArguments>(process.argv);

/**
 * Probability threshold for considering given difference signficiant.
 */
const PROBABILITY_CONSIDERED_SIGNIFICANT = 0.02;

/**
 * Probability threshold for considering given difference meaningless.
 */
const PROBABILITY_CONSIDERED_MEANINGLESS = 0.05;

/**
 * Render duration threshold (in ms) for treating given difference as significant.
 *
 * This is additional filter, in addition to probability threshold above.
 * Too small duration difference might be result of measurement grain of 1 ms.
 */
const DURATION_DIFF_THRESHOLD_SIGNIFICANT = 4;

/**
 * Render duration threshold (in ms) for treating given difference as meaningless.
 *
 * This is an overriding filter that can override probability threshold above.
 * Too small duration difference might be result of measurement grain of 1 ms.
 */
const DURATION_DIFF_THRESHOLD_MININGLESS = 2;

const errors: string[] = [];
const warnings: string[] = [];

function logError(message: string, ...args: any[]) {
  errors.push(message);
  console.error(`ðŸ›‘ ${message}`, ...args);
}

function logWarning(message: string) {
  warnings.push(message);
  console.warn(`ðŸŸ¡ ${message}`);
}

/**
 * Main routine.
 *
 * Responsible for loading baseline and current performance results and outputting data in various formats.
 */
export async function main() {
  try {
    const hasCurrentFile = fsSync.existsSync(currentFilePath);
    if (!hasCurrentFile) {
      logError(`Current results files "${currentFilePath}" does not exists. Check your setup.`);
      process.exit(1);
    }

    const current = await loadFile(currentFilePath);

    const hasBaslineFile = fsSync.existsSync(baselineFilePath);
    if (!hasBaslineFile) {
      logWarning(
        `Baseline results files "${baselineFilePath}" does not exists. This warning should be ignored only if you are bootstapping perf test setup, otherwise it indicates invalid setup.`
      );
    }

    const baseline = hasBaslineFile ? await loadFile(baselineFilePath) : null;

    const outputData = compareResults(current, baseline);

    if (output === 'console' || output === 'all') printToConsole(outputData);
    if (output === 'json' || output === 'all') writeToJson(outputFilePath, outputData);
    if (output === 'markdown' || output === 'all') writeToMarkdown('compare-output.md', outputData);
  } catch (error) {
    logError(`Error.`, error);
    process.exit(1);
  }
}

/**
 * Load performance file and parse it to `PerformanceRecord` object.
 */
async function loadFile(path: string): Promise<PerformanceRecord> {
  const data = await fs.readFile(path, 'utf8');

  const lines = data.split(/\r?\n/);
  // TODO: add data format validation
  const entries: PerformanceEntry[] = lines.filter((line) => !!line.trim()).map((line) => JSON.parse(line));

  if (hasDuplicateValues(entries.map((entry) => entry.name))) {
    const msg = `Your performance result file ${path} contains records with duplicated names.
      Please remove any non-unique names from your test suites and try again.
      `;
    logError(msg);
    throw new Error(msg);
  }

  const result: PerformanceRecord = {};
  entries.forEach((entry) => {
    result[entry.name] = entry;
  });

  return result;
}

/**
 * Compare results between baseline and current entries and categorize.
 */
function compareResults(currentEntries: PerformanceRecord, baselineEntries: PerformanceRecord | null): CompareResult {
  // unique keys
  const names = [...new Set([...Object.keys(currentEntries), ...Object.keys(baselineEntries || {})])];

  const regular: CompareEntry[] = [];
  const added: AddedEntry[] = [];
  const removed: RemovedEntry[] = [];

  names.forEach((name) => {
    const current = currentEntries[name];
    const baseline = baselineEntries?.[name];

    if (baseline && current) {
      regular.push(buildCompareEntry(name, current, baseline));
    } else if (current) {
      added.push({ name, current });
    } else if (baseline) {
      removed.push({ name, baseline });
    }
  });

  const significant = regular
    .filter((item) => item.durationDiffSignificance === 'SIGNIFICANT')
    .sort((a, b) => b.durationDiff - a.durationDiff);
  const insignificant = regular
    .filter((item) => item.durationDiffSignificance === 'INSIGNIFICANT')
    .sort((a, b) => b.durationDiff - a.durationDiff);
  const meaningless = regular
    .filter((item) => item.durationDiffSignificance === 'MEANINGLESS')
    .sort((a, b) => b.durationDiff - a.durationDiff);
  const countChanged = regular.filter((item) => item.countDiff !== 0).sort((a, b) => b.countDiff - a.countDiff);
  added.sort((a, b) => b.current.meanDuration - a.current.meanDuration);
  removed.sort((a, b) => b.baseline.meanDuration - a.baseline.meanDuration);

  return {
    significant,
    insignificant,
    meaningless,
    countChanged,
    added,
    removed,
    errors,
    warnings,
  };
}

/**
 * Establish statisticial signficance of render duration difference build compare entry.
 */
function buildCompareEntry(name: string, current: PerformanceEntry, baseline: PerformanceEntry): CompareEntry {
  const durationDiff = current.meanDuration - baseline.meanDuration;
  const relativeDurationDiff = durationDiff / baseline.meanDuration;
  const countDiff = current.meanCount - baseline.meanCount;
  const relativeCountDiff = countDiff / baseline.meanCount;

  const z = computeZ(baseline.meanDuration, baseline.stdevDuration, current.meanDuration, current.runs);
  const prob = computeProbability(z);

  let durationDiffSignificance: StatisticalSignificance = 'INSIGNIFICANT';
  if (prob < PROBABILITY_CONSIDERED_SIGNIFICANT && Math.abs(durationDiff) >= DURATION_DIFF_THRESHOLD_SIGNIFICANT)
    durationDiffSignificance = 'SIGNIFICANT';

  if (prob > PROBABILITY_CONSIDERED_MEANINGLESS || Math.abs(durationDiff) <= DURATION_DIFF_THRESHOLD_MININGLESS)
    durationDiffSignificance = 'MEANINGLESS';

  return {
    name,
    baseline,
    current,
    durationDiff,
    relativeDurationDiff,
    durationDiffSignificance,
    countDiff,
    relativeCountDiff,
  };
}

/**
 * Calculate z-score for given baseline and current performance results.
 *
 * Based on :: https://github.com/v8/v8/blob/master/test/benchmarks/csuite/compare-baseline.py
 */
function computeZ(baselineMean: number, baselineStdev: number, currentMean: number, runs: number) {
  if (baselineStdev == 0) return 1000;

  return Math.abs((currentMean - baselineMean) / (baselineStdev / Math.sqrt(runs)));
}

/**
 * Compute statisticial hyphotesis probability based on z-score.
 *
 * Based on :: https://github.com/v8/v8/blob/master/test/benchmarks/csuite/compare-baseline.py
 */
function computeProbability(z: number) {
  // p 0.005: two sided < 0.01
  if (z > 2.575_829) return 0;

  // p 0.010
  if (z > 2.326_348) return 0.01;

  // p 0.015
  if (z > 2.170_091) return 0.02;

  // p 0.020
  if (z > 2.053_749) return 0.03;

  // p 0.025: two sided < 0.05
  if (z > 1.959_964) return 0.04;

  // p 0.030
  if (z > 1.880_793) return 0.05;

  // p 0.035
  if (z > 1.811_91) return 0.06;

  // p 0.040
  if (z > 1.750_686) return 0.07;

  // p 0.045
  if (z > 1.695_397) return 0.08;

  // p 0.050: two sided < 0.10
  if (z > 1.644_853) return 0.09;

  // p 0.100: two sided < 0.20
  if (z > 1.281_551) return 0.1;

  // two sided p >= 0.20
  return 0.2;
}

main();
