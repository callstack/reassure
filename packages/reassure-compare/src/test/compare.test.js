import { loadFile } from '../compare';

describe('loadFile', () => {
  it('should load results file with header', async () => {
    const results = loadFile(`${__dirname}/valid-header.perf`);

    expect(results.metadata).toEqual({
      branch: 'feat/perf-file-validation',
      commitHash: '991427a413b1ff05497a881287c9ddcba7b8de54',
    });

    const entries = Object.keys(results.entries);
    expect(entries).toHaveLength(5);
    expect(entries).toEqual([
      'Other Component 10',
      'Other Component 10 legacy scenario',
      'Other Component 20',
      'Async Component',
      'fib 30',
    ]);
    expect(results).toMatchSnapshot();
  });

  it('should load results file without header', async () => {
    const results = loadFile(`${__dirname}/valid-no-header.perf`);

    expect(results.metadata).toBeUndefined();

    const entries = Object.keys(results.entries);
    expect(entries).toHaveLength(5);
    expect(entries).toEqual([
      'Other Component 10',
      'Other Component 10 legacy scenario',
      'Other Component 20',
      'Async Component',
      'fib 30',
    ]);
    expect(results).toMatchSnapshot();
  });

  it('should fail for file with invalid JSON structure', async () => {
    expect(() => loadFile(`${__dirname}/invalid-json.perf`)).toThrowErrorMatchingSnapshot();
  });

  it('should fail for file with invalid entry', async () => {
    expect(() => loadFile(`${__dirname}/invalid-entry.perf`)).toThrowErrorMatchingSnapshot();
  });

  it('should support entries without type', async () => {
    const results = loadFile(`${__dirname}/default-type.perf`);

    const types = Object.entries(results.entries).map(([key, value]) => value.type);
    expect(types).toEqual(['render', 'render', 'function']);
  });
});
