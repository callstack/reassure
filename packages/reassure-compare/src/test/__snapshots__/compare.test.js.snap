// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadFile should fail for file with invalid JSON structure 1`] = `"Unexpected token : in JSON at position 6"`;

exports[`loadFile should fail for file with invalid entry 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      1,
      "runs"
    ],
    "message": "Expected number, received string"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      1,
      "meanDuration"
    ],
    "message": "Required"
  }
]"
`;

exports[`loadFile should load results file with header 1`] = `
{
  "entries": {
    "Async Component": {
      "counts": [
        7,
        7,
        7,
        7,
        7,
        7,
        7,
        7,
        7,
        7,
      ],
      "durations": [
        160,
        158,
        158,
        155,
        153,
        150,
        149,
        145,
        144,
        136,
      ],
      "meanCount": 7,
      "meanDuration": 150.8,
      "name": "Async Component",
      "runs": 10,
      "stdevCount": 0,
      "stdevDuration": 7.554248252914823,
      "type": "render",
    },
    "Other Component 10": {
      "counts": [
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
      ],
      "durations": [
        100,
        97,
        95,
        94,
        94,
        94,
        93,
        90,
        86,
        84,
      ],
      "meanCount": 4,
      "meanDuration": 92.7,
      "name": "Other Component 10",
      "runs": 10,
      "stdevCount": 0,
      "stdevDuration": 4.831608887776871,
      "type": "render",
    },
    "Other Component 10 legacy scenario": {
      "counts": [
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
      ],
      "durations": [
        97,
        96,
        94,
        92,
        91,
        91,
        88,
        88,
        85,
        83,
      ],
      "meanCount": 4,
      "meanDuration": 90.5,
      "name": "Other Component 10 legacy scenario",
      "runs": 10,
      "stdevCount": 0,
      "stdevDuration": 4.552166761249221,
      "type": "render",
    },
    "Other Component 20": {
      "counts": [
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
      ],
      "durations": [
        99,
        99,
        98,
        98,
        96,
        95,
        94,
        93,
        93,
        93,
        93,
        92,
        90,
        90,
        89,
        88,
        88,
        87,
        87,
        82,
      ],
      "meanCount": 4,
      "meanDuration": 92.2,
      "name": "Other Component 20",
      "runs": 20,
      "stdevCount": 0,
      "stdevDuration": 4.595191995301633,
      "type": "render",
    },
    "fib 30": {
      "counts": [
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
      ],
      "durations": [
        80,
        80,
        80,
        80,
        80,
        79,
        79,
        79,
        79,
        79,
      ],
      "meanCount": 1,
      "meanDuration": 79.9,
      "name": "fib 30",
      "runs": 10,
      "stdevCount": 0,
      "stdevDuration": 0.4532780026900862,
      "type": "function",
    },
  },
  "metadata": {
    "branch": "feat/perf-file-validation",
    "commitHash": "991427a413b1ff05497a881287c9ddcba7b8de54",
  },
}
`;

exports[`loadFile should load results file without header 1`] = `
{
  "entries": {
    "Async Component": {
      "counts": [
        7,
        7,
        7,
        7,
        7,
        7,
        7,
        7,
        7,
        7,
      ],
      "durations": [
        160,
        158,
        158,
        155,
        153,
        150,
        149,
        145,
        144,
        136,
      ],
      "meanCount": 7,
      "meanDuration": 150.8,
      "name": "Async Component",
      "runs": 10,
      "stdevCount": 0,
      "stdevDuration": 7.554248252914823,
      "type": "render",
    },
    "Other Component 10": {
      "counts": [
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
      ],
      "durations": [
        100,
        97,
        95,
        94,
        94,
        94,
        93,
        90,
        86,
        84,
      ],
      "meanCount": 4,
      "meanDuration": 92.7,
      "name": "Other Component 10",
      "runs": 10,
      "stdevCount": 0,
      "stdevDuration": 4.831608887776871,
      "type": "render",
    },
    "Other Component 10 legacy scenario": {
      "counts": [
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
      ],
      "durations": [
        97,
        96,
        94,
        92,
        91,
        91,
        88,
        88,
        85,
        83,
      ],
      "meanCount": 4,
      "meanDuration": 90.5,
      "name": "Other Component 10 legacy scenario",
      "runs": 10,
      "stdevCount": 0,
      "stdevDuration": 4.552166761249221,
      "type": "render",
    },
    "Other Component 20": {
      "counts": [
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
      ],
      "durations": [
        99,
        99,
        98,
        98,
        96,
        95,
        94,
        93,
        93,
        93,
        93,
        92,
        90,
        90,
        89,
        88,
        88,
        87,
        87,
        82,
      ],
      "meanCount": 4,
      "meanDuration": 92.2,
      "name": "Other Component 20",
      "runs": 20,
      "stdevCount": 0,
      "stdevDuration": 4.595191995301633,
      "type": "render",
    },
    "fib 30": {
      "counts": [
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
        1,
      ],
      "durations": [
        80,
        80,
        80,
        80,
        80,
        79,
        79,
        79,
        79,
        79,
      ],
      "meanCount": 1,
      "meanDuration": 79.9,
      "name": "fib 30",
      "runs": 10,
      "stdevCount": 0,
      "stdevDuration": 0.4532780026900862,
      "type": "function",
    },
  },
  "metadata": undefined,
}
`;
